Build fails on Windows: linker chokes due to missing function InitSecurityInterfaceA

My build is failing on Windows, with static-php-cli 2.6.1.

Here's the build log: https://github.com/drupal/cms-launcher/actions/runs/16333991122/job/46142404600#step:8:5704

The error that takes it down seems to be (including the backtrace):

```
Creating library D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\php.lib and object D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\php.exp
libcurl_a.lib(curl_sspi.obj) : error LNK2001: unresolved external symbol InitSecurityInterfaceA
D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\php.exe : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: '"link.exe" /nologo @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\PHP_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\CLI_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\BZ2_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\CTYPE_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\CURL_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\DATE_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\FILTER_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\GD_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\HASH_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\ICONV_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\JSON_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\MBSTRING_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\OPCACHE_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\OPENSSL_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\PCRE_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\RANDOM_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\REFLECTION_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\SESSION_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\SPL_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\SQLITE3_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\STANDARD_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\TOKENIZER_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\YAML_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\ZIP_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\ZLIB_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\LIBXML_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\DOM_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\PDO_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\PDO_SQLITE_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\PHAR_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\SIMPLEXML_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\XML_GLOBAL_OBJS.txt" @"D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\resp\XMLWRITER_GLOBAL_OBJS.txt" libbz2_a.lib  libcurl_a.lib libcrypto.lib libssl.lib Crypt32.Lib WinMM.Lib Wldap32.Lib normaliz.lib libssh2.lib nghttp2.lib   zlib_a.lib libjpeg_a.lib libpng_a.lib libfreetype_a.lib libiconv_a.lib libwebp.lib libsharpyuv.lib avif.lib User32.Lib Gdi32.Lib  libiconv_a.lib    libcrypto.lib libssl.lib Crypt32.Lib      libsqlite3_a.lib iphlpapi.lib  yaml.lib libzip_a.lib libbz2_a.lib zlib_a.lib liblzma_a.lib zlib_a.lib libxml2_a.lib libiconv_a.lib   libsqlite3_a.lib     D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\Zend\jump_x86_64_ms_pe_masm.obj D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\Zend\make_x86_64_ms_pe_masm.obj kernel32.lib ole32.lib user32.lib advapi32.lib shell32.lib ws2_32.lib Dnsapi.lib psapi.lib bcrypt.lib ws2_32.lib shell32.lib D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\php.exe.res /out:D:\a\cms-launcher\cms-launcher\source\php-src\x64\Release\php.exe /nologo /libpath:"D:\a\cms-launcher\cms-launcher\buildroot\lib" /stack:67108864 /ltcg /nodefaultlib:msvcrt /nodefaultlib:msvcrtd /ignore:4286' : return code '0x460'
Stop.
[01:35:28] [E] Uncaught SPC\exception\RuntimeException: Command run failed with code[2]: cd /d "D:\a\cms-launcher\cms-launcher\source\php-src" && D:\a\cms-launcher\cms-launcher\php-sdk-binary-tools\phpsdk-vs17-x64.bat -t nmake_cli_wrapper.bat --task-args php.exe at D:\a\cms-launcher\cms-launcher\src\globals\functions.php(177)
[01:35:28] [E] #0 D:\a\cms-launcher\cms-launcher\src\SPC\util\WindowsCmd.php(52): f_passthru()
#1 D:\a\cms-launcher\cms-launcher\src\SPC\builder\windows\WindowsBuilder.php(167): SPC\util\WindowsCmd->exec()
#2 D:\a\cms-launcher\cms-launcher\src\SPC\builder\windows\WindowsBuilder.php(133): SPC\builder\windows\WindowsBuilder->buildCli()
#3 D:\a\cms-launcher\cms-launcher\src\SPC\command\BuildPHPCommand.php(208): SPC\builder\windows\WindowsBuilder->buildPHP()
#4 D:\a\cms-launcher\cms-launcher\src\SPC\command\BaseCommand.php(107): SPC\command\BuildPHPCommand->handle()
#5 D:\a\cms-launcher\cms-launcher\vendor\symfony\console\Command\Command.php(326): SPC\command\BaseCommand->execute()
#6 D:\a\cms-launcher\cms-launcher\vendor\symfony\console\Application.php(1078): Symfony\Component\Console\Command\Command->run()
#7 D:\a\cms-launcher\cms-launcher\vendor\symfony\console\Application.php(324): Symfony\Component\Console\Application->doRunCommand()
#8 D:\a\cms-launcher\cms-launcher\vendor\symfony\console\Application.php(175): Symfony\Component\Console\Application->doRun()
#9 D:\a\cms-launcher\cms-launcher\bin\spc(31): Symfony\Component\Console\Application->run()
#10 {main}
```

A known issue? I haven't changed _anything_ about the build process part from bumping static-php-cli to 2.6.1, but this was also happening on 2.5.2.

--- a/config/env.ini
+++ b/config/env.ini
@@ -64,7 +64,7 @@ UPX_EXEC="${PKG_ROOT_PATH}\bin\upx.exe"
 ; phpmicro patches, for more info, see: https://github.com/easysoft/phpmicro/tree/master/patches
 SPC_MICRO_PATCHES=static_extensions_win32,cli_checks,disable_huge_page,vcruntime140,win32,zend_stream,cli_static
 ; Windows static linking system libs
-SPC_WINDOWS_EXEC_LIBS="ws2_32.lib shell32.lib"
+SPC_EXTRA_LIBS=""
 
 [linux]
 ; Linux can use different build toolchain, but the toolchain can not be changed in this file:

--- a/config/env.ini
+++ b/config/env.ini
@@ -63,6 +63,8 @@ PHP_SDK_PATH="${WORKING_DIR}\php-sdk-binary-tools"
 UPX_EXEC="${PKG_ROOT_PATH}\bin\upx.exe"
 ; phpmicro patches, for more info, see: https://github.com/easysoft/phpmicro/tree/master/patches
 SPC_MICRO_PATCHES=static_extensions_win32,cli_checks,disable_huge_page,vcruntime140,win32,zend_stream,cli_static
+; Windows static linking system libs
+SPC_WINDOWS_EXEC_LIBS="ws2_32.lib shell32.lib"
 
 [linux]
 ; Linux can use different build toolchain, but the toolchain can not be changed in this file:

--- a/config/source.json
+++ b/config/source.json
@@ -4,7 +4,8 @@
         "license": {
             "type": "file",
             "path": "LICENSE"
-        }
+        },
+        "alt": false
     },
     "amqp": {
         "type": "url",

--- a/src/SPC/builder/extension/curl.php
+++ b/src/SPC/builder/extension/curl.php
@@ -59,10 +59,10 @@ class curl extends Extension
 
     public function patchBeforeMake(): bool
     {
-        $extra_libs = getenv('SPC_WINDOWS_EXEC_LIBS');
+        $extra_libs = getenv('SPC_EXTRA_LIBS');
         if (!str_contains($extra_libs, 'secur32.lib')) {
             $extra_libs .= ' secur32.lib';
-            putenv('SPC_WINDOWS_EXEC_LIBS=' . $extra_libs);
+            putenv('SPC_EXTRA_LIBS=' . trim($extra_libs));
             return true;
         }
         return false;

--- a/src/SPC/builder/extension/curl.php
+++ b/src/SPC/builder/extension/curl.php
@@ -57,6 +57,17 @@ class curl extends Extension
         return true;
     }
 
+    public function patchBeforeMake(): bool
+    {
+        $extra_libs = getenv('SPC_WINDOWS_EXEC_LIBS');
+        if (!str_contains($extra_libs, 'secur32.lib')) {
+            $extra_libs .= ' secur32.lib';
+            putenv('SPC_WINDOWS_EXEC_LIBS=' . $extra_libs);
+            return true;
+        }
+        return false;
+    }
+
     public function patchBeforeSharedConfigure(): bool
     {
         $file = $this->source_dir . '/config.m4';

--- a/src/SPC/builder/windows/WindowsBuilder.php
+++ b/src/SPC/builder/windows/WindowsBuilder.php
@@ -162,8 +162,10 @@ class WindowsBuilder extends BuilderBase
     {
         SourcePatcher::patchWindowsCLITarget();
 
+        $extra_libs = getenv('SPC_WINDOWS_EXEC_LIBS') ?: '';
+
         // add nmake wrapper
-        FileSystem::writeFile(SOURCE_PATH . '\php-src\nmake_cli_wrapper.bat', "nmake /nologo LIBS_CLI=\"{$this->getOption('extra-libs')} ws2_32.lib shell32.lib\" EXTRA_LD_FLAGS_PROGRAM= %*");
+        FileSystem::writeFile(SOURCE_PATH . '\php-src\nmake_cli_wrapper.bat', "nmake /nologo LIBS_CLI=\"{$extra_libs}\" EXTRA_LD_FLAGS_PROGRAM= %*");
 
         cmd()->cd(SOURCE_PATH . '\php-src')->exec("{$this->sdk_prefix} nmake_cli_wrapper.bat --task-args php.exe");
 
@@ -197,9 +199,11 @@ class WindowsBuilder extends BuilderBase
         }
         FileSystem::writeFile(SOURCE_PATH . '\php-src\Makefile', $makefile);
 
+        $extra_libs = getenv('SPC_WINDOWS_EXEC_LIBS') ?: '';
+
         // add nmake wrapper
         $fake_cli = $this->getOption('with-micro-fake-cli', false) ? ' /DPHP_MICRO_FAKE_CLI" ' : '';
-        $wrapper = "nmake /nologo LIBS_MICRO=\"{$this->getOption('extra-libs')} ws2_32.lib shell32.lib\" CFLAGS_MICRO=\"/DZEND_ENABLE_STATIC_TSRMLS_CACHE=1{$fake_cli}\" %*";
+        $wrapper = "nmake /nologo LIBS_MICRO=\"{$extra_libs}\" CFLAGS_MICRO=\"/DZEND_ENABLE_STATIC_TSRMLS_CACHE=1{$fake_cli}\" %*";
         FileSystem::writeFile(SOURCE_PATH . '\php-src\nmake_micro_wrapper.bat', $wrapper);
 
         // phar patch for micro

--- a/src/SPC/builder/windows/WindowsBuilder.php
+++ b/src/SPC/builder/windows/WindowsBuilder.php
@@ -61,11 +61,6 @@ class WindowsBuilder extends BuilderBase
      */
     public function buildPHP(int $build_target = BUILD_TARGET_NONE): void
     {
-        // ---------- Update extra-libs ----------
-        $extra_libs = getenv('SPC_EXTRA_LIBS') ?: '';
-        $extra_libs .= (empty($extra_libs) ? '' : ' ') . implode(' ', $this->getAllStaticLibFiles());
-        f_putenv('SPC_EXTRA_LIBS=' . $extra_libs);
-
         $enableCli = ($build_target & BUILD_TARGET_CLI) === BUILD_TARGET_CLI;
         $enableFpm = ($build_target & BUILD_TARGET_FPM) === BUILD_TARGET_FPM;
         $enableMicro = ($build_target & BUILD_TARGET_MICRO) === BUILD_TARGET_MICRO;
@@ -162,10 +157,10 @@ class WindowsBuilder extends BuilderBase
     {
         SourcePatcher::patchWindowsCLITarget();
 
-        $extra_libs = getenv('SPC_WINDOWS_EXEC_LIBS') ?: '';
+        $extra_libs = getenv('SPC_EXTRA_LIBS') ?: '';
 
         // add nmake wrapper
-        FileSystem::writeFile(SOURCE_PATH . '\php-src\nmake_cli_wrapper.bat', "nmake /nologo LIBS_CLI=\"{$extra_libs}\" EXTRA_LD_FLAGS_PROGRAM= %*");
+        FileSystem::writeFile(SOURCE_PATH . '\php-src\nmake_cli_wrapper.bat', "nmake /nologo LIBS_CLI=\"ws2_32.lib shell32.lib {$extra_libs}\" EXTRA_LD_FLAGS_PROGRAM= %*");
 
         cmd()->cd(SOURCE_PATH . '\php-src')->exec("{$this->sdk_prefix} nmake_cli_wrapper.bat --task-args php.exe");
 
@@ -199,11 +194,11 @@ class WindowsBuilder extends BuilderBase
         }
         FileSystem::writeFile(SOURCE_PATH . '\php-src\Makefile', $makefile);
 
-        $extra_libs = getenv('SPC_WINDOWS_EXEC_LIBS') ?: '';
+        $extra_libs = getenv('SPC_EXTRA_LIBS') ?: '';
 
         // add nmake wrapper
         $fake_cli = $this->getOption('with-micro-fake-cli', false) ? ' /DPHP_MICRO_FAKE_CLI" ' : '';
-        $wrapper = "nmake /nologo LIBS_MICRO=\"{$extra_libs}\" CFLAGS_MICRO=\"/DZEND_ENABLE_STATIC_TSRMLS_CACHE=1{$fake_cli}\" %*";
+        $wrapper = "nmake /nologo LIBS_MICRO=\"ws2_32.lib shell32.lib {$extra_libs}\" CFLAGS_MICRO=\"/DZEND_ENABLE_STATIC_TSRMLS_CACHE=1{$fake_cli}\" %*";
         FileSystem::writeFile(SOURCE_PATH . '\php-src\nmake_micro_wrapper.bat', $wrapper);
 
         // phar patch for micro

--- a/src/SPC/builder/windows/library/curl.php
+++ b/src/SPC/builder/windows/library/curl.php
@@ -37,6 +37,7 @@ class curl extends WindowsLibraryBase
                 '-DBUILD_EXAMPLES=OFF ' . // disable examples
                 '-DUSE_LIBIDN2=OFF ' . // disable libidn2
                 '-DCURL_USE_LIBPSL=OFF ' . // disable libpsl
+                '-DUSE_WINDOWS_SSPI=ON ' . // use Schannel instead of OpenSSL
                 '-DCURL_USE_SCHANNEL=ON ' . // use Schannel instead of OpenSSL
                 '-DCURL_USE_OPENSSL=OFF ' . // disable openssl due to certificate issue
                 '-DCURL_ENABLE_SSL=ON ' .

--- a/src/SPC/command/DownloadCommand.php
+++ b/src/SPC/command/DownloadCommand.php
@@ -252,7 +252,7 @@ class DownloadCommand extends BaseCommand
                         } elseif ($alt_sources === false) {
                             logger()->warning("No alternative sources found for {$source}, skipping alternative download");
                             throw $e;
-                        } else{
+                        } else {
                             logger()->notice("Trying to download alternative sources for {$source}");
                             $alt_config = array_merge($config, $alt_sources);
                         }

--- a/src/SPC/command/DownloadCommand.php
+++ b/src/SPC/command/DownloadCommand.php
@@ -245,11 +245,15 @@ class DownloadCommand extends BaseCommand
                         }
                         // if download failed, we will try to download alternative sources
                         logger()->warning("Download failed: {$e->getMessage()}");
-                        logger()->notice("Trying to download alternative sources for {$source}");
                         $alt_sources = Config::getSource($source)['alt'] ?? null;
                         if ($alt_sources === null) {
+                            logger()->warning("No alternative sources found for {$source}, using default alternative source");
                             $alt_config = array_merge($config, $this->getDefaultAlternativeSource($source));
-                        } else {
+                        } elseif ($alt_sources === false) {
+                            logger()->warning("No alternative sources found for {$source}, skipping alternative download");
+                            throw $e;
+                        } else{
+                            logger()->notice("Trying to download alternative sources for {$source}");
                             $alt_config = array_merge($config, $alt_sources);
                         }
                         Downloader::downloadSource($source, $alt_config, $force_all || in_array($source, $force_list));

--- a/src/globals/test-extensions.php
+++ b/src/globals/test-extensions.php
@@ -24,12 +24,12 @@ $test_os = [
     // 'macos-13', // bin/spc for x86_64
     // 'macos-14',  // bin/spc for arm64
     // 'macos-15', // bin/spc for arm64
-    'ubuntu-latest', // bin/spc-alpine-docker for x86_64
+    // 'ubuntu-latest', // bin/spc-alpine-docker for x86_64
     // 'ubuntu-22.04', // bin/spc-gnu-docker for x86_64
     // 'ubuntu-24.04', // bin/spc for x86_64
     // 'ubuntu-22.04-arm', // bin/spc-gnu-docker for arm64
     // 'ubuntu-24.04-arm', // bin/spc for arm64
-    // 'windows-latest', // .\bin\spc.ps1
+    'windows-latest', // .\bin\spc.ps1
 ];
 
 // whether enable thread safe
@@ -49,7 +49,7 @@ $prefer_pre_built = false;
 // If you want to test your added extensions and libs, add below (comma separated, example `bcmath,openssl`).
 $extensions = match (PHP_OS_FAMILY) {
     'Linux', 'Darwin' => 'imap,swoole',
-    'Windows' => 'intl',
+    'Windows' => 'curl',
 };
 
 // If you want to test shared extensions, add them below (comma separated, example `bcmath,openssl`).

